File: LeetCode/BinaryTreeDFS/python/flattenBinaryTree.py

from collections import deque
from typing import Dict, List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None:
            return root
        
        self.DFS(root)
    
    def DFS(self, root: Optional[TreeNode]) -> TreeNode:
        if root is None:
            return root
        # The order is necessary. We need to DFS on the left child first. Then DFS on the right child
        leftTail = self.DFS(root.left)
        rightTail = self.DFS(root.right)

        # Each time we DFS, we check for the left child
        # We need to connect the right child of the root, to the left tail 
        # Then, assight the left child to the right child
        # Lastly, set left child to null
        if root.left is not None:
            leftTail.right = root.right
            root.right = root.left
            root.left = None

        # Use boolean and it needs to be in this order. Compiler process OR statements left to right
        last = rightTail or leftTail or root

        return last


    @staticmethod
    def testSolution(input: Dict[str, List[str|int]]) -> None:
        # Create the Binary Tree from the input list using BFS
        binTree = Solution.BFSCreateBinaryTree(input["root"])
        expectedTree = Solution.BFSCreateBinaryTree(input["expected"])

        # Print the Binary Tree and the expected output
        print("Input: " + Solution.BFSBinaryTreeToStr(binTree))
        print("Expected: " + Solution.BFSBinaryTreeToStr(expectedTree))

        # Call the function to be tested
        Solution().flatten(binTree)
   
        # Print the result
        print("Result: " + Solution.BFSBinaryTreeToStr(binTree))

        # Validate the result
        print("PASS" if Solution.validateResult(binTree, expectedTree) else "FAIL")
        print("-" * 50)

    @staticmethod
    def BFSCreateBinaryTree(node: List[int|str]) -> TreeNode:
        arLen = len(node)

        if arLen < 1:
            return None

        root = TreeNode(node[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            # Pre order: root, left, then right
            cur_node = queue.popleft()

            # Left child
            if i < arLen and node[i] != "null":
                cur_node.left = TreeNode(node[i], None, None)
                queue.append(cur_node.left)
            i += 1
            
            # Right child
            if i < arLen and node[i] != "null":
                cur_node.right = TreeNode(node[i], None, None)
                queue.append(cur_node.right)            
            i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[TreeNode]) -> None:

        if root is None:
            return "[]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[" + ", ".join(sb) + "]"

    @staticmethod
    def validateResult(res: Optional[TreeNode], expected: Optional[TreeNode]) -> bool:
        resList = []
        expectedList = []

        qRes = deque([res])
        qExpected = deque([expected])

        while len(qRes) > 0:
            curNode = qRes.popleft()

            resList.append(curNode.val if curNode is not None else "null")

            if curNode is not None:
                qRes.append(curNode.left)
                qRes.append(curNode.right)

        while len(qExpected) > 0:
            curNode = qExpected.popleft()
            
            expectedList.append(curNode.val if curNode is not None else "null")

            if curNode is not None:
                qExpected.append(curNode.left)
                qExpected.append(curNode.right)

        # print(resList)
        # print(expectedList)

        return resList == expectedList


if __name__ == "__main__":
    input = {"root": [1,2,5,3,4,"null",6], "expected": [1,"null",2,"null",3,"null",4,"null",5,"null",6]}
    Solution.testSolution(input)

    input = {"root": [], "expected": []}
    Solution.testSolution(input)

    input = {"root": [0], "expected": [0]}
    Solution.testSolution(input)

File: LeetCode/BinaryTreeDFS/java/flattenBinaryTree/flattenBinaryTree.java
package LeetCode.BinaryTreeDFS.java.flattenBinaryTree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

record flattenBinaryTreeRecord(String[] root, String[] expected) {
}

class flattenBinaryTree {
    public void flatten(TreeNode root) {
        this.DFS(root);
    }

    private TreeNode DFS(TreeNode root) {
        if (root == null) {
            return root;
        }

        // The order is necessary. We need to DFS on the left child first. Then DFS on
        // the right child
        TreeNode leftTail = this.DFS(root.left);
        TreeNode rightTail = this.DFS(root.right);

        // Each time we DFS, we check for the left child
        // We need to connect the right child of the root, to the left tail
        // Then, assight the left child to the right child
        // Lastly, set left child to null
        if (root.left != null) {
            leftTail.right = root.right;
            root.right = root.left;
            root.left = null;
        }

        // Use boolean and it needs to be in this order. Compiler process OR statements
        // left to right
        TreeNode last = rightTail != null ? rightTail : (leftTail != null ? leftTail : root);

        return last;
    }

    public static void main(String[] args) {
        flattenBinaryTreeRecord input = new flattenBinaryTreeRecord(
                new String[] { "1", "2", "5", "3", "4", "null", "6" },
                new String[] { "1", "null", "2", "null", "3", "null", "4", "null", "5", "null", "6" });
        flattenBinaryTree.testSolution(input);

        input = new flattenBinaryTreeRecord(
                new String[] {},
                new String[] {});
        flattenBinaryTree.testSolution(input);

        input = new flattenBinaryTreeRecord(
                new String[] { "0" },
                new String[] { "0" });
        flattenBinaryTree.testSolution(input);
    }

    private static void testSolution(flattenBinaryTreeRecord input) {
        // Create the binary tree from the input array using BFS
        TreeNode rootBTree = flattenBinaryTree.BFSCreateBinaryTree(input.root());
        TreeNode expdBTree = flattenBinaryTree.BFSCreateBinaryTree(input.expected());

        // Print the binary tree and the expected output
        System.out.println("Input: root: " + flattenBinaryTree.BFSBinaryTreeToStr(rootBTree));
        System.out.println("Expected: " + flattenBinaryTree.BFSBinaryTreeToStr(expdBTree));

        // Call the function to be tested
        new flattenBinaryTree().flatten(rootBTree);

        // Print the result
        System.out.println("Result: " + flattenBinaryTree.BFSBinaryTreeToStr(rootBTree));

        // Validate the result
        System.out.println(flattenBinaryTree.validateResult(rootBTree, expdBTree) ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }

    private static TreeNode BFSCreateBinaryTree(String[] nodeValsArray) {
        int arLen = nodeValsArray.length;

        if (arLen < 1) {
            return null;
        }

        TreeNode root = new TreeNode(Integer.parseInt(nodeValsArray[0]), null, null);
        int i = 1;

        // Using queue (BFS) to build the tree level by level
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.addLast(root);

        while (!queue.isEmpty() && i < arLen) {
            TreeNode curNode = queue.removeFirst();

            if (curNode != null) {
                // left child
                if (i < arLen && nodeValsArray[i] != "null") {
                    curNode.left = new TreeNode(Integer.parseInt(nodeValsArray[i]), null, null);
                    queue.addLast(curNode.left);
                }
                i++;

                // right child
                if (i < arLen && nodeValsArray[i] != "null") {
                    curNode.right = new TreeNode(Integer.parseInt(nodeValsArray[i]), null, null);
                    queue.addLast(curNode.right);
                }
                i++;
            }
        }

        return root;
    }

    private static String BFSBinaryTreeToStr(TreeNode root) {
        // Using BFS traversal to print tree value level by level
        if (root == null) {
            return "[]";
        }

        List<String> sb = new ArrayList<>();

        Deque<TreeNode> queue = new LinkedList<>(); // Use linkedList to allow addition of null object. ArrayDeque does
                                                    // not allow null object
        queue.addLast(root);

        while (!queue.isEmpty()) {
            TreeNode curNode = queue.removeFirst();
            sb.add(curNode != null ? String.valueOf(curNode.val) : "null");

            if (curNode != null) {
                queue.addLast(curNode.left);
                queue.addLast(curNode.right);
            }
        }

        while (!sb.isEmpty()) {
            if (sb.getLast() == "null") {
                sb.removeLast();
            } else {
                break;
            }
        }

        return sb.toString();
    }

    private static boolean validateResult(TreeNode res, TreeNode expectedTree) {
        List<String> resList = new ArrayList<>();
        List<String> expectedList = new ArrayList<>();

        // Use linkedList to allow addition of null object. ArrayDeque does
        Deque<TreeNode> qRes = new LinkedList<>();
        qRes.addLast(res);

        Deque<TreeNode> qExpected = new LinkedList<>();
        qExpected.addLast(expectedTree);

        while (!qRes.isEmpty()) {
            TreeNode curNode = qRes.pollFirst();

            resList.add(curNode != null ? String.valueOf(curNode.val) : "null");

            if (curNode != null) {
                qRes.addLast(curNode.left);
                qRes.addLast(curNode.right);
            }
        }

        while (!qExpected.isEmpty()) {
            TreeNode curNode = qExpected.pollFirst();

            expectedList.add(curNode != null ? String.valueOf(curNode.val) : "null");

            if (curNode != null) {
                qExpected.addLast(curNode.left);
                qExpected.addLast(curNode.right);
            }
        }

        return expectedList.equals(resList);
    }
}
